File: 940-distinct-subsequences-ii/940-distinct-subsequences-ii.cpp
class Solution {
public:
    int distinctSubseqII(string s) {
        int res = 0, added = 0, mod = 1e9 + 7, endsWith[26] = {0};
        for(auto c : s){
            added = (res + 1 - endsWith[c - 'a']) % mod;
            res = (res + added) % mod;
            endsWith[c - 'a'] = (endsWith[c - 'a'] + added) % mod;
        }
        return (res + mod) % mod;
    }
};

File: 940-distinct-subsequences-ii/940-distinct-subsequences-ii.java
class Solution {
    public int distinctSubseqII(String s) {
        int end[] = new int[26], res = 0, added = 0, mod = (int)1e9 + 7;
        for (char c : s.toCharArray()){
            added = (res + 1 - end[c - 'a']) % mod;
            res = (res + added) % mod;
            end[c - 'a'] = (end[c - 'a'] + added) % mod;
        }
        return (res + mod) % mod;
    }
}